### DISCUSSION TITLE
Conditional validation based on key to gain performance

### DISCUSSION THREAD

**Discussion by NixBiks** [pydantic/pydantic#2919] [Question]
Imagine a `Component` model that is a `Union` of heaps of `BaseModel`s. Each with an `id`

```python
from pydantic import BaseModel
from typing_extensions import Literal
from typing import Union


class A(BaseModel):
  id: Literal["a"]
  name: str


class B(BaseModel):
  id: Literal["b"]
  value: float


class Component(BaseModel):
  __root__: Union[A, B]
    

Component.parse_obj({"id": "b", "value": 123})
```

The issue with is that `Component.parse_obj` becomes very expensive due to the `Union` (there can be nested models etc.). I imagine this can be improved a lot by exploiting the fact that the `id` simply reveals which model should be used for validation. E.g. if `id` is `a` then `A.parse_obj` should be used.

I'm not sure how to enforce this conditional validation with pydantic though. One approach could be to define `__init__` on `Component` and then somehow call the specific models `__init__` instead of `super().__init__` - I'm not sure whether that's possible though. 

Another approach could be to use `root_validator` where I can validate and parse the data but it doesn't really give the performance boost I'm looking for (if any!?)

```python
class Component(BaseModel):
  __root__: Union[A, B]

  @root_validator(pre=True)
  def validate(cls, values):
    d = values.get("__root__")
    if d.get("id") == "a":
      return {"__root__": A.parse_obj(d)}
    elif d.get("id") == "b":
      return {"__root__": B.parse_obj(d)}
    
    return values

Component.parse_obj({"id": "b", "value": 123})
```

**Comment by PrettyWood** [comment]
Hi @mr-bjerre 
https://github.com/samuelcolvin/pydantic/pull/2336 ;)
